### ISSUE TITLE
kernel memory consumption regression due to open inodes in snapshots

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#16539]
Since: 5e55954f278f2f823e4fe3edaf7600f78d2922f9

Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After 5e55954f278f2f823e4fe3edaf7600f78d2922f9, it rises to ~7k when taking a snapshot, since we open those files for reading.

Kernel memory is extra precious since we only reserve a small amount for the kernel.

**Comment by raphaelsc** [comment]
> Since: [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9)
> 
> Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9), it rises to ~7k when taking a snapshot, since we open those files for reading.
> 
> Kernel memory is extra precious since we only reserve a small amount for the kernel.

Tablets are relatively small in size and migrating tablets is a small fraction of all (the ones that would cause increase in overhead). Load balancer defines a per-shard limit (4 IIRC) on streaming read concurrency. So it's 4 * #_of_sstables_per_tablet (can be significant with LCS or large partitions, but let's assume the former. 10G / 0.16 = 62 sstables) * #_of_shards. Up to 23808 streaming sstables in a node with 96 shards = ~170M of memory. Significant. And we could fix the problem without the overhead, so let's do it.

**Comment by avikivity** [comment]
> > Since: [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9)
> > Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9), it rises to ~7k when taking a snapshot, since we open those files for reading.
> > Kernel memory is extra precious since we only reserve a small amount for the kernel.
> 
> Tablets are relatively small in size and migrating tablets is a small fraction of all (the ones that would cause increase in overhead). Load balancer defines a per-shard limit (4 IIRC) on streaming read concurrency. So it's 4 * #_of_sstables_per_tablet (can be significant with LCS or large partitions, but let's assume the former. 10G / 0.16 = 62 sstables) * #_of_shards. Up to 23808 streaming sstables in a node with 96 shards = ~170M of memory. Significant. And we could fix the problem without the overhead, so let's do it.

170MB is not significant if you have 96 shards.

I forgot that only a few tablets are active at any time. So I don't think there's a problem there.

What about a full snapshot?

**Comment by raphaelsc** [comment]
> > > Since: [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9)
> > > Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9), it rises to ~7k when taking a snapshot, since we open those files for reading.
> > > Kernel memory is extra precious since we only reserve a small amount for the kernel.
> > 
> > 
> > Tablets are relatively small in size and migrating tablets is a small fraction of all (the ones that would cause increase in overhead). Load balancer defines a per-shard limit (4 IIRC) on streaming read concurrency. So it's 4 * #_of_sstables_per_tablet (can be significant with LCS or large partitions, but let's assume the former. 10G / 0.16 = 62 sstables) * #_of_shards. Up to 23808 streaming sstables in a node with 96 shards = ~170M of memory. Significant. And we could fix the problem without the overhead, so let's do it.
> 
> 170MB is not significant if you have 96 shards.
> 
> I forgot that only a few tablets are active at any time. So I don't think there's a problem there.
> 
> What about a full snapshot?

The snapshot today is used in the context of file streaming, but it's better to no make assumptions about its usage. Let's say tomorrow we find a use case for it where we want to save a snapshot of all tablets to s3, right?

**Comment by avikivity** [comment]
> > > > Since: [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9)
> > > > Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9), it rises to ~7k when taking a snapshot, since we open those files for reading.
> > > > Kernel memory is extra precious since we only reserve a small amount for the kernel.
> > > 
> > > 
> > > Tablets are relatively small in size and migrating tablets is a small fraction of all (the ones that would cause increase in overhead). Load balancer defines a per-shard limit (4 IIRC) on streaming read concurrency. So it's 4 * #_of_sstables_per_tablet (can be significant with LCS or large partitions, but let's assume the former. 10G / 0.16 = 62 sstables) * #_of_shards. Up to 23808 streaming sstables in a node with 96 shards = ~170M of memory. Significant. And we could fix the problem without the overhead, so let's do it.
> > 
> > 
> > 170MB is not significant if you have 96 shards.
> > I forgot that only a few tablets are active at any time. So I don't think there's a problem there.
> > What about a full snapshot?
> 
> The snapshot today is used in the context of file streaming, but it's better to no make assumptions about its usage. Let's say tomorrow we find a use case for it where we want to save a snapshot of all tablets to s3, right?

What about regular snapshot? from `nodetool snapshot`? Does it not use it?

You're right about hardening it for the future, but it decreases the priority.

**Comment by raphaelsc** [comment]
> > > > > Since: [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9)
> > > > > Right now an sstable's footprint in kernel memory is about ~2k, for the data and index inodes. After [5e55954](https://github.com/scylladb/scylladb/commit/5e55954f278f2f823e4fe3edaf7600f78d2922f9), it rises to ~7k when taking a snapshot, since we open those files for reading.
> > > > > Kernel memory is extra precious since we only reserve a small amount for the kernel.
> > > > 
> > > > 
> > > > Tablets are relatively small in size and migrating tablets is a small fraction of all (the ones that would cause increase in overhead). Load balancer defines a per-shard limit (4 IIRC) on streaming read concurrency. So it's 4 * #_of_sstables_per_tablet (can be significant with LCS or large partitions, but let's assume the former. 10G / 0.16 = 62 sstables) * #_of_shards. Up to 23808 streaming sstables in a node with 96 shards = ~170M of memory. Significant. And we could fix the problem without the overhead, so let's do it.
> > > 
> > > 
> > > 170MB is not significant if you have 96 shards.
> > > I forgot that only a few tablets are active at any time. So I don't think there's a problem there.
> > > What about a full snapshot?
> > 
> > 
> > The snapshot today is used in the context of file streaming, but it's better to no make assumptions about its usage. Let's say tomorrow we find a use case for it where we want to save a snapshot of all tablets to s3, right?
> 
> What about regular snapshot? from `nodetool snapshot`? Does it not use it?
> 
> You're right about hardening it for the future, but it decreases the priority.

No, the usage of that procedure, today, is restricted to file streaming. I should probably even make the interface more explicit about that by switching to tablet ids rather than token range.
